python/triton/ops/flash_attention.py

```diff
@@ -24,34 +24,29 @@ def _fwd_kernel(
     stride_vz, stride_vh, stride_vn, stride_vk,
     stride_oz, stride_oh, stride_om, stride_on,
     Z, H, N_CTX,
+    Z_H_N_CTX,
     BLOCK_M: tl.constexpr, BLOCK_DMODEL: tl.constexpr,
     BLOCK_N: tl.constexpr,
     IS_CAUSAL: tl.constexpr,
 ):
     start_m = tl.program_id(0)
     off_hz = tl.program_id(1)
     qvk_offset = off_hz * stride_qh
-    Q_block_ptr = tl.make_block_ptr(
-        base=Q + qvk_offset,
-        shape=(N_CTX, BLOCK_DMODEL),
-        strides=(stride_qm, stride_qk),
-        offsets=(start_m * BLOCK_M, 0),
-        block_shape=(BLOCK_M, BLOCK_DMODEL),
-        order=(1, 0)
-    )
+    vk_offset = qvk_offset // stride_qm
+
     K_block_ptr = tl.make_block_ptr(
-        base=K + qvk_offset,
-        shape=(BLOCK_DMODEL, N_CTX),
+        base=K,
```

Comment:
@ptillet need your attention for comments:
1. Currently, we have a restriction that basePtr/shape/stride must be directly from kernel arguments
2. We are working on an enhancement to enable that basePtr/shape/stride indirectly from scalar kernel arguments, not in very high priority though.
3. We cannot easily support basePtr/shape/stride from other than 2, like in this case, the basePtr is from getProgramId(). This is currently not in the plan, it can be very heavy unless some change from cuda driver.

Reply:
yeah, I think that's fair. Doesn't seem like a big deal to rewrite the kernel this way

Reply:
Hi @ptillet @goostavz  , it is little difficult for me to understand why you changed Q_block_ptr construction to this:

```
# K is not transposed with shape (z, h, N_CTX/*128*/, d/*64*/) and strides (h*N_CTX*d, N_CTX*d/*8192*/, d/*64*/, 1)

K.shape == (emb_d/*64*/,  Z*H*N_CTX)
```

instead of original

```
K.shape == (emb_d/*64*/, N_CTX)
```

In each cuda block (Z*H * N_CTX/BLOCK_SIZE_N), should we load Z*H*N_CTX emb_d vectors ?

It looks like deep related to LLVM dialect optimization related to TMA. Could you explain to me what happened behind this ?

Reply:
B.t.w, since this op is developed for compute ability both 8.0 and 9.0 (Hopper), and there is no TMA available for card of 8.0. I think checking maybe desired ?